### PR TITLE
Cpp tutorial

### PR DESCRIPTION
**Pull request by renefritze**

### CODE REVIEW

**Comment by HenKlei** [comment]
@renefritze Just a short question since you added a number for the tutorial in 4bfaaa0: I am preparing a PR to rename `tutorial01` and remove the number. I think that was the last agreement in #929 and during the latest dev-meeting, wasn't it?

**Comment by renefritze** [comment]
Maybe I misremember, but didn't we want to do that after the release?

**Comment by HenKlei** [comment]
According to the notes from the last meeting, the numbers should be removed for this release and finding a good ordering/structure should happen after the release.

**Comment by sdrave** [comment]
@HenKlei, let's wait with the cleanup until all tutorials have been merged.

**Comment by HenKlei** [comment]
> @HenKlei, let's wait with the cleanup until all tutorials have been merged.

Alright. Then we can also standardize the naming of the `.rst` tutorial files.

**Comment by renefritze** [comment]
Please go and have a look now @pymor/pymor-devs . 
Some notes:
  1. The `OneDVisualizer` plots are not showing up in docs yet. 
  1. I did not get the `meth` links for `axpy,scal` etc working. Could someone please tell me what I'm doing wrong?
  1. There's currently a bug in jupyter-sphinx that sets the wrong filename on bash kernel script downloads, hence the somewhat weird phrasing.

**Comment by renefritze** [comment]
1. depends on https://github.com/pymor/pymor/pull/966

**Comment by sdrave** [comment]
Shouldn't https://docs.pymor.org/cpp_tutorial be working?

**Comment by renefritze** [comment]
Hmm, there's some `_` `-` auto replace in gitlab refname at work  I think:
https://docs.pymor.org/cpp-tutorial/tutorial_external_solver.html

File: docs/source/tutorial_external_solver.rst [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+|VectorArray|, |VectorSpace| and |Operator| that
+
+Instead of writing a full implementaion of a |VectorArray| that manages multiple `model.Vector`
+instances we can instead implement a wrapper `WrappedVector` for a single `model.Vector` instance based on
+:class:`~pymor.vectorarrays.list.CopyOnWriteVector` which will be used to create
+|ListVectorArray| via a :class:`~pymor.vectorarrays.list.ListVectorSpace` based `WrappedVectorSpace`.
+
+The :class:`~pymor.vectorarrays.list.CopyOnWriteVector` base class manages a reference count for
+us and automatically copies data when necessary in methods :meth:`~pymor.vectorarrays.list.CopyOnWriteVector.scal`
+and :meth:`~pymor.vectorarrays.list.CopyOnWriteVector.axpy`. To use this we need to implement
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:171]:
```suggestion
and :meth:`~pymor.vectorarrays.list.CopyOnWriteVector.axpy`. To use this, we need to implement
```

File: docs/source/tutorial_external_solver.rst [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+
+      def dofs(self, dof_indices):
+          raise NotImplementedError
+
+      def amax(self):
+          raise NotImplementedError
+
+
+The implementation of the `WrappedVectorSpace` is very short as most of the necessary methods
+of |VectorSpace| are default implemented in :class:`~pymor.vectorarrays.list.ListVectorSpace`.
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:238]:
```suggestion
of |VectorSpace| are implemented in :class:`~pymor.vectorarrays.list.ListVectorSpace`.
```

File: docs/source/tutorial_external_solver.rst [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+
+      def make_vector(self, obj):
+          # obj is a `model.Vector` instance
+          return WrappedVector(obj)
+
+      def __eq__(self, other):
+          return type(other) is WrappedVectorSpace and self.dim == other.dim
+
+
+Wrapping the `model.DiffusionOperator` is straight forward as well. We just need to attach
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:258]:
```suggestion
Wrapping the `model.DiffusionOperator` is straightforward as well. We just need to attach
```

File: docs/source/tutorial_external_solver.rst [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+              self.op.apply(u._impl, v)
+              return v
+
+          return self.range.make_array([apply_one_vector(u) for u in U._list])
+
+
+Putting it all together
+-----------------------
+
+As a demonstration we will use our (stationary) toy diffusion solver in a transient example with timestepping
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:290]:
```suggestion
As a demonstration, we will use our (stationary) toy diffusion solver in a transient example with timestepping
```

File: docs/source/tutorial_external_solver.rst [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+provided by pyMOR. For details on the discretization setup see :doc:`tutorial01`. First up we implement
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:291]:
```suggestion
provided by pyMOR. For details on the discretization setup, see :doc:`tutorial01`. First up, we implement
```

File: docs/source/tutorial_external_solver.rst [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+  def discretize(n, nt, blocks):
+      h = 1. / blocks
+      ops = [WrappedDiffusionOperator.create(n, h * i, h * (i + 1)) for i in range(blocks)]
+      pfs = [ProjectionParameterFunctional('diffusion_coefficients', blocks, i) for i in range(blocks)]
+      operator = LincombOperator(ops, pfs)
+
+      initial_data = operator.source.zeros()
+
+      # use to_numpy method of WrappedVector to setup rhs
+      # note that we cannot use the data property of ListVectorArray,
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:316]: The `data` property is deprecated.

File: docs/source/tutorial_external_solver.rst [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+
+      time_stepper = ExplicitEulerTimeStepper(nt)
+
+      fom = InstationaryModel(T=1e-0, operator=operator, rhs=rhs, initial_data=initial_data,
+                              time_stepper=time_stepper, num_values=20,
+                              visualizer=visualizer, name='C++-Model')
+      return fom
+
+Now we can build a reduced basis for our model. Note that this code is not specific to our wrapped classes. Those
+only directly used in the `discretize` call.
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:337]: Something is missing in the last sentence.

File: docs/source/tutorial_external_solver.rst [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+      rhs_data[len(rhs_data) - 1] = 0
+      rhs = VectorOperator(rhs_vec)
+
+      # we can re-use pyMOR's builtin grid and visualizer for our demonstration
+      grid = OnedGrid(domain=(0, 1), num_intervals=n)
+      visualizer = OnedVisualizer(grid)
+
+      time_stepper = ExplicitEulerTimeStepper(nt)
+
+      fom = InstationaryModel(T=1e-0, operator=operator, rhs=rhs, initial_data=initial_data,
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:331]:
```suggestion
      fom = InstationaryModel(T=1, operator=operator, rhs=rhs, initial_data=initial_data,
```

File: docs/source/tutorial_external_solver.rst [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+  fom = discretize(50, 10000, 4)
+  parameter_space = fom.parameters.space(0.1, 1)
+
+  # generate solution snapshots
+  snapshots = fom.solution_space.empty()
+  for mu in parameter_space.sample_uniformly(2):
+      snapshots.append(fom.solve(mu))
+
+  # apply POD
+  reduced_basis = pod(snapshots, modes=4)[0]
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:352]: This creates two snapshots, but asks for 4 POD modes?

File: docs/source/tutorial_external_solver.rst [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+
+  # reduce the model
+  reductor = InstationaryRBReductor(fom, reduced_basis, check_orthonormality=True)
+  rom = reductor.reduce()
+
+  # stochastic error estimation
+  mu_max = None
+  err_max = -1.
+  for mu in parameter_space.sample_randomly(10):
+      U_RB = (reductor.reconstruct(rom.solve(mu)))
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:362]:
```suggestion
      U_RB = reductor.reconstruct(rom.solve(mu))
```

File: docs/source/tutorial_external_solver.rst [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+      if err > err_max:
+          err_max = err
+          mu_max = mu
+
+  # visualize maximum error solution
+  U_RB = (reductor.reconstruct(rom.solve(mu_max)))
+  U = fom.solve(mu_max)
+  fom.visualize((U_RB, U), title=f'mu = {mu}', legend=('reduced', 'detailed'))
+
+As you can see in this comparison we get a good approximation of the full order model here and
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:374]:
```suggestion
As you can see in this comparison, we get a good approximation of the full-order model here and
```

File: docs/source/tutorial_external_solver.rst [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+the error plot confirms it:
+
+.. jupyter-execute::
+
+  fom.visualize((U-U_RB), title=f'mu = {mu}', legend=('error'))
+
+You can download this demonstration plus the wrapper definitions as a
+notebook :jupyter-download:notebook:`tutorial_external_solver` or
+as a plain python script :jupyter-download:script:`tutorial_external_solver`.
```
Comment (pmli) [inline docs/source/tutorial_external_solver.rst:383]:
```suggestion
as a plain Python script :jupyter-download:script:`tutorial_external_solver`.
```

**Comment by pmli** [comment]
> 2. I did not get the `meth` links for `axpy,scal` etc working. Could someone please tell me what I'm doing wrong?

The issue seems to be that the methods of `Vector` and `CopyOnWriteVector` are not documented. Adding a docstring to `CopyOnWriteVector.scal` makes the link work.

**Comment by renefritze** [comment]
> For some reason, `visualize` does not produce any plots on my machine.

Thanks for the input @pmli. I've noted this in 2. The OneDVisualizer currently wants to pop up a QT window, hence #966

File: docs/source/tutorial_external_solver.rst [hunk 13]
```diff
@@ -0,0 +1,383 @@
+Tutorial: Binding external PDE solver to pyMOR
+================================================
```
[the remark this answers is not in the document]
Comment (renefritze) [reply docs/source/tutorial_external_solver.rst:2]: This was just nicer, not necessary, right?

File: docs/source/tutorial_external_solver.rst [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+
+
+Using the exported Python classes with pyMOR
+--------------------------------------------
+
+..
+  This only works after merging https://github.com/pymor/pymor/pull/1013
+  In :doc:`tutorial-rb`
+
+In another tutorial we have learned how pyMOR's algorithms operate on |VectorArrays| and |Operators|. To be able to use
```
[the remark this answers is not in the document]
Comment (renefritze) [reply docs/source/tutorial_external_solver.rst:160]: Didn't consider that about Operators.

File: docs/source/tutorial_external_solver.rst (hunk 6 above)
Comment (renefritze) [reply docs/source/tutorial_external_solver.rst:316]: Hah, caught in the first, but not the second line :roll_eyes:

File: docs/source/tutorial_external_solver.rst (hunk 7 above)
Comment (renefritze) [reply docs/source/tutorial_external_solver.rst:337]: 'Those wrapped types are only directly used' maybe?

File: docs/source/tutorial_external_solver.rst (hunk 9 above)
Comment (renefritze) [reply docs/source/tutorial_external_solver.rst:352]: Tbh, I didn't much look at the code. This is straight c&p form the existing demo

**Comment by renefritze** [comment]
> > 1. I did not get the `meth` links for `axpy,scal` etc working. Could someone please tell me what I'm doing wrong?
> 
> The issue seems to be that the methods of `Vector` and `CopyOnWriteVector` are not documented. Adding a docstring to `CopyOnWriteVector.scal` makes the link work.

D'oh. Thanks! Somehow I thought sphinx would warn if I'm linking against something that's not there and therefore assumed I was doing sth wrong.

File: docs/source/tutorial_external_solver.rst (hunk 7 above)
Comment (ftalbrecht) [inline docs/source/tutorial_external_solver.rst]: This sentence is missing something

File: docs/source/tutorial_external_solver.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (pmli) [reply docs/source/tutorial_external_solver.rst]: Right.

File: docs/source/tutorial_external_solver.rst (hunk 7 above)
Comment (pmli) [reply docs/source/tutorial_external_solver.rst]: Sure.

File: docs/source/tutorial_external_solver.rst (hunk 9 above)
Comment (pmli) [reply docs/source/tutorial_external_solver.rst:351]: Ah, sorry, it's an instationary model...

**Comment by pmli** [comment]
> > For some reason, `visualize` does not produce any plots on my machine.
> 
> Thanks for the input @pmli. I've noted this in 1. The OneDVisualizer currently wants to pop up a QT window, hence #966

Aha, ok, I was wondering why there were two plots that popped up when running `make docs`.

> Somehow I thought sphinx would warn if I'm linking against something that's not there and therefore assumed I was doing sth wrong.

Yes, it's strange. Should we address the issue of documenting `Vector` in another PR?

**Comment by renefritze** [comment]
> > Somehow I thought sphinx would warn if I'm linking against something that's not there and therefore assumed I was doing sth wrong.
> 
> Should we address the issue of documenting `Vector` in another PR?

Sounds good, yeah.

**Comment by renefritze** [comment]
This actually depends on first merging #966  and #1009 and afterwards using the method form #1009 in #966.

File: docs/source/tutorial_external_solver.rst [hunk 15]
```diff
@@ -0,0 +1,382 @@
+Tutorial: Binding an external PDE solver to pyMOR
+=================================================
+
+One of pyMOR's main features is easy integration of external solvers. In this tutorial
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:4]:
```suggestion
One of pyMOR's main features is easy integration of external solvers that implement the full-order model. In this tutorial
```

File: docs/source/tutorial_external_solver.rst [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+we will do this step-by-step for a toy solver of the one-dimensional diffusion equation written in C++.
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:5]:
```suggestion
we will do this step-by-step for a custom toy solver of the one-dimensional diffusion equation written in C++.
If you use the `FEniCS <https://fenicsproject.org>`_ or `NGSovle <https://ngsolve.org>`_ PDE solver libraries,
you can find ready-to-use pyMOR bindings in the :mod:`~pymor.bindings` package. pyMOR support for
`deal.II <https://dealii.org>`_ can be found in a `separate repository <https://github.com/pymor/pymor-deal.II>`_.
```

File: docs/source/tutorial_external_solver.rst [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+
+
+Defining the PDE solver
+-----------------------
+
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:10]: Briefly define the equation that is solved and say how it is discretized?

File: docs/source/tutorial_external_solver.rst [hunk 18]
```diff
@@ -0,0 +1,382 @@
...
+Just like the diffusion operator that computes a central differences stencil:
+
+.. literalinclude:: minimal_cpp_demo/model.cc
+    :lines: 39-49
+    :language: cpp
+
+This completes all the C++ code needed for the toy solver itself. Next, we will make this code usable from Python.
+We utilize the `pybind11 <https://github.com/pybind/pybind11>`_ library to create a Python
+`extension module <https://docs.python.org/3/extending/extending.html>`_ named `model`, that allows us to manipulate
+instances of the C++ `Vector` and `DiffusionOperator` classes.
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:44]:
```suggestion
instances of the C++ `Vector` and `DiffusionOperator` classes.

Compiling the PDE solver as a shared library and creating Python bindings for it using 
`pybind11 <https://github.com/pybind/pybind11>`_, `Cython <https://cython.org>`_ or
`ctypes <https://docs.python.org/3/library/ctypes.html>`_ is the preferred way of integrating
external solvers, as it offers maximal flexibility and performance. For instance, in this
example we will actually completely implement the |Model| in Python using a
:mod:`time stepper <pymor.algorithms.timestepping>` from pyMOR to
:meth:`~pymor.models.interface.Model.solve` the |Model|.

When this is not an option,
`RPC <https://en.wikipedia.org/wiki/Remote_procedure_call>`-based approaches are
possible as well. For small to medium-sized linear problems, another option it to import
system matrices and snapshot data into pyMOR via file exchange and to use NumPy-based
:mod:`Operators <pymor.operators.numpy>` and :mod:`VectorArrays <pymor.vectorarrays.numpy>`
to represent the full-order model.
```

File: docs/source/tutorial_external_solver.rst [hunk 19]
```diff
@@ -0,0 +1,382 @@
...
+    :lines: 60-61
+    :language: cpp
+
+Next, we define read-only properties on the Python side named after and delegated to the members of the C++ class.
+
+.. literalinclude:: minimal_cpp_demo/model.cc
+    :lines: 62-63
+    :language: cpp
+
+And the last `DiffusionOperator`-related line exposes the function call to apply in the same way:
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:71]:
```suggestion
The last `DiffusionOperator`-related line exposes the function call to apply in the same way:
```

File: docs/source/tutorial_external_solver.rst [hunk 20]
```diff
@@ -0,0 +1,382 @@
...
+with a few more advanced techniques added.
+
+.. literalinclude:: minimal_cpp_demo/model.cc
+    :lines: 66-68
+    :language: cpp
+
+Again we define a `py:class\_` with appropiate name and docstring, but now we also indicate to pybind11
+that this class will implement the `buffer protocol <https://docs.python.org/3/c-api/buffer.html>`_, which basically
+exposes direct access to the chunk of memory associated with a `Vector` instance to Python. We also see how we can dispatch multiple init functions
+by using `py:init` objects with C++ lambda functions.
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:87]:
```suggestion
by using `py:init` objects with C++ lambda functions.
Note that direct memory access to the vector data from Python is not required to integrate a solver with pyMOR.
It is, however, useful for debugging and quickly modifying or extending the solver from within Python. For instance,
in our toy example we will use the direct memory access to quickly define a visualization of the solutions and to
construct the right-hand side vector for our problem.
```

File: docs/source/tutorial_external_solver.rst [hunk 21]
```diff
@@ -0,0 +1,382 @@
...
+
+
+Using the exported Python classes with pyMOR
+--------------------------------------------
+
+..
+  This only works after merging https://github.com/pymor/pymor/pull/1013
+  In :doc:`tutorial-rb`
+
+In another tutorial, we have learned how pyMOR's algorithms operate on |VectorArrays| and |Operators|. To be able to use
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:159]:
```suggestion
All of pyMOR's algorithms operate on |VectorArray| and |Operator| objects that all share the same programming interface. To be able to use
```

File: docs/source/tutorial_external_solver.rst [hunk 22]
```diff
@@ -0,0 +1,382 @@
...
+our Python `model.Vector` and `model.DiffusionOperator` in pyMOR, we have to provide implementations of
+|VectorArray|, |VectorSpace| and |Operator| that wrap the classes defined in the extension module.
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:161]:
```suggestion
|VectorArray|, |VectorSpace| and |Operator| that wrap the classes defined in the extension module
and translate calls to the interface methods into operations on `model.Vector` and `model.DiffusionOperator`.
```

File: docs/source/tutorial_external_solver.rst [hunk 23]
```diff
@@ -0,0 +1,382 @@
...
+
+Instead of writing a full implementaion of a |VectorArray| that manages multiple `model.Vector`
+instances, we can instead implement a wrapper `WrappedVector` for a single `model.Vector` instance based on
+:class:`~pymor.vectorarrays.list.CopyOnWriteVector` which will be used to create
+|ListVectorArray| via a :class:`~pymor.vectorarrays.list.ListVectorSpace`-based `WrappedVectorSpace`.
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:166]:
```suggestion
|ListVectorArrays| via a :class:`~pymor.vectorarrays.list.ListVectorSpace`-based `WrappedVectorSpace`.
```

File: docs/source/tutorial_external_solver.rst [hunk 24]
```diff
@@ -0,0 +1,382 @@
...
+              self.op.apply(u._impl, v)
+              return v
+
+          return self.range.make_array([apply_one_vector(u) for u in U._list])
+
+
+Putting it all together
+-----------------------
+
+As a demonstration, we will use our (stationary) toy diffusion solver in a transient example with timestepping
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:289]:
```suggestion
As a demonstration, we will use our toy diffusion solver in a transient example with timestepping
```

File: docs/source/tutorial_external_solver.rst [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+provided by pyMOR. For details on the discretization setup, see :doc:`tutorial01`. First up, we implement
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:290]:
```suggestion
provided by pyMOR. First up, we implement
```

I don't think that this tutorial really is a good reference, as much is said that isn't needed here and many other things are missing.

File: docs/source/tutorial_external_solver.rst [hunk 26]
```diff
@@ -0,0 +1,382 @@
...
+      pfs = [ProjectionParameterFunctional('diffusion_coefficients', blocks, i) for i in range(blocks)]
+      operator = LincombOperator(ops, pfs)
+
+      initial_data = operator.source.zeros()
+
+      # use to_numpy method of WrappedVector to setup rhs
+      # note that we cannot use the to_numpy method of ListVectorArray,
+      # since ListVectorArray will always return a copy
+      rhs_vec = operator.range.zeros()
+      rhs_data = rhs_vec._list[0].to_numpy()
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:318]: This is a bit outdated. The following should work as well:

```suggestion
      rhs_vec = operator.range.zeros()
      rhs_data = rhs_vec._data[0]
```

**Comment by renefritze** [comment]
The visualization will work after #1017. Otherwise this is now good to merge.

File: docs/source/tutorial_external_solver.rst [hunk 27]
```diff
@@ -0,0 +1,412 @@
...
+Compiling the PDE solver as a shared library and creating Python bindings for it using
+`pybind11 <https://github.com/pybind/pybind11>`_, `Cython <https://cython.org>`_ or
+`ctypes <https://docs.python.org/3/library/ctypes.html>`_ is the preferred way of integrating
+external solvers, as it offers maximal flexibility and performance. For instance, in this
+example we will actually completely implement the |Model| in Python using a
+:mod:`time stepper <pymor.algorithms.timestepping>` from pyMOR to
+:meth:`~pymor.models.interface.Model.solve` the |Model|.
+
+When this is not an option,
+`RPC <https://en.wikipedia.org/wiki/Remote_procedure_call>`-based approaches are
```
Comment (sdrave) [inline docs/source/tutorial_external_solver.rst:60]:
```suggestion
`RPC <https://en.wikipedia.org/wiki/Remote_procedure_call>`_-based approaches are
```

**Pull request merged**